OtripleS.Web.Api/Controllers/TeacherContactsController.cs

```diff
@@ -0,0 +1,156 @@
+﻿// ---------------------------------------------------------------
+// Copyright (c) Coalition of the Good-Hearted Engineers
+// FREE TO USE AS LONG AS SOFTWARE FUNDS ARE DONATED TO THE POOR
+// ---------------------------------------------------------------
+
+using Microsoft.AspNetCore.Mvc;
+using OtripleS.Web.Api.Models.TeacherContacts;
+using OtripleS.Web.Api.Models.TeacherContacts.Exceptions;
+using OtripleS.Web.Api.Services.TeacherContacts;
+using RESTFulSense.Controllers;
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace OtripleS.Web.Api.Controllers
+{
+	[Route("api/[controller]")]
+	[ApiController]
+	public class TeacherContactsController : RESTFulController
+	{
+		private readonly ITeacherContactService teacherContactService;
+
+		public TeacherContactsController(ITeacherContactService teacherContactService) =>
+			this.teacherContactService = teacherContactService;
+
+		[HttpPost]
+		public async ValueTask<ActionResult<TeacherContact>> PostTeacherContactAsync(
+			TeacherContact teacherContact)
+		{
+			try
+			{
+				TeacherContact persistedTeacherContact =
+					await this.teacherContactService.AddTeacherContactAsync(teacherContact);
+
+				return Ok(persistedTeacherContact);
+			}
+			catch (TeacherContactValidationException teacherContactValidationException)
+				when (teacherContactValidationException.InnerException is AlreadyExistsTeacherContactException)
+			{
+				string innerMessage = GetInnerMessage(teacherContactValidationException);
+
+				return Conflict(innerMessage);
+			}
+			catch (TeacherContactValidationException teacherContactValidationException)
+			{
+				string innerMessage = GetInnerMessage(teacherContactValidationException);
+
+				return BadRequest(innerMessage);
+			}
+			catch (TeacherContactDependencyException teacherContactDependencyException)
+			{
+				return Problem(teacherContactDependencyException.Message);
+			}
+			catch (TeacherContactServiceException teacherContactServiceException)
+			{
+				return Problem(teacherContactServiceException.Message);
+			}
+		}
+
+		[HttpGet]
+		public ActionResult<IQueryable<TeacherContact>> GetAllTeacherContacts()
+		{
+			try
+			{
+				IQueryable storageTeacherContacts =
+					this.teacherContactService.RetrieveAllTeacherContacts();
+
+				return Ok(storageTeacherContacts);
+			}
+			catch (TeacherContactDependencyException teacherContactDependencyException)
+			{
+				return Problem(teacherContactDependencyException.Message);
+			}
+			catch (TeacherContactServiceException teacherContactServiceException)
+			{
+				return Problem(teacherContactServiceException.Message);
+			}
+		}
+
+		[HttpGet("teacher/{teacherId}/contact/{contactId}")]
```

Comment:
this should be `teachers/{teacherId}/contacts/{contactId}` plural singular plural singular please